src/hotspot/share/jfr/periodic/sampling/jfrCPUTimeThreadSampler.cpp

```diff
@@ -0,0 +1,1041 @@
+    jtl->unset_timerid();
+    jtl->cpu_time_jfr_stack().ensure_capacity(0);
+    s4 lost_samples = jtl->get_and_reset_cpu_time_lost_samples();
+    if (lost_samples > 0) {
+      JfrCPUTimeThreadSampler::record_lost_event(thread, lost_samples);
+    }
+  }
+}
+
+void JfrCPUTimeThreadSampler::start_thread() {
+  if (os::create_thread(this, os::os_thread)) {
+    os::start_thread(this);
+  } else {
+    log_error(jfr)("Failed to create thread for thread sampling");
+  }
+}
+
+void JfrCPUTimeThreadSampler::enroll() {
+  if (Atomic::cmpxchg(&_disenrolled, true, false)) {
+    log_info(jfr)("Enrolling CPU thread sampler");
+    _sample.signal();
+    init_timers();
+    log_trace(jfr)("Enrolled CPU thread sampler");
+  }
+}
+
+void JfrCPUTimeThreadSampler::disenroll() {
+  if (!Atomic::cmpxchg(&_disenrolled, false, true)) {
+    log_info(jfr)("Disenrolling CPU thread sampler");
+    stop_timer();
+    Atomic::store(&_stop_signals, true);
+    while (_active_signal_handlers > 0) {
+      // wait for all signal handlers to finish
+      os::naked_short_nanosleep(1000);
+    }
+    _sample.wait();
+    Atomic::store(&_stop_signals, false);
+    log_trace(jfr)("Disenrolled CPU thread sampler");
+  }
+}
+
+void JfrCPUTimeThreadSampler::run() {
+  assert(_sampler_thread == nullptr, "invariant");
+  _sampler_thread = this;
+  int64_t last_autoadapt_check = os::javaTimeNanos();
+  while (true) {
+    if (!_sample.trywait()) {
+      // disenrolled
+      _sample.wait();
+    }
+    _sample.signal();
+
+
+    if (os::javaTimeNanos() - last_autoadapt_check > AUTOADAPT_INTERVAL_MS * 1000000) {
+      autoadapt_period_if_needed();
+      last_autoadapt_check = os::javaTimeNanos();
+    }
+
+    if (is_out_of_safepoint_recording_triggered()) {
+      untrigger_out_of_safepoint_recording();
+      record_out_of_safepoint();
+    }
+    os::naked_sleep(100);
+  }
+}
+
+void JfrCPUTimeThreadSampler::record_out_of_safepoint() {
+  ResourceMark rm;
+  MutexLocker tlock(Threads_lock);
+  ThreadsListHandle tlh;
+  for (size_t i = 0; i < tlh.list()->length(); i++) {
+    JavaThread* jt = tlh.list()->thread_at(i);
+    JfrThreadLocal* jtl = jt->jfr_thread_local();
+    if (jtl == nullptr) {
+      continue;
+    }
+    if (jtl->wants_cpu_time_out_of_safepoint_recording()) {
+      if (!jtl->acquire_cpu_time_jfr_out_of_safepoint_lock()) {
+        continue;
+      }
+      jtl->set_wants_cpu_time_out_of_safepoint_recording(false);
+      record_out_of_safepoint(jt, jtl);
+      jtl->release_cpu_time_jfr_queue_lock();
+    }
+  }
+}
+
+
+void JfrCPUTimeThreadSampler::record_out_of_safepoint(JavaThread* thread, JfrThreadLocal* jtl) {
+  JfrLocalCPUTimeTraceStack& stack = jtl->cpu_time_jfr_stack();
+  JfrStackFrame* tmp_stackframes = JfrCHeapObj::new_array<JfrStackFrame>(_max_frames_per_trace);
+ // assert(!stack.is_empty(), "invariant");
+  for (u4 i = 0; i < stack.size(); i++) {
+    JfrCPUTimeThreadSampler::record_event(thread, stack.at(i), tmp_stackframes);
```

Comment:
Shouldn’t we return the trace to the _frame_store as with on_safepoint()?

Reply:
Good catch.